internal/querycoord/task_scheduler.go

```diff
@@ -608,53 +603,61 @@ func (scheduler *TaskScheduler) scheduleLoop() {
 		triggerTaskKeys = append(triggerTaskKeys, taskKey)
 		triggerTaskKeys = append(triggerTaskKeys, stateKey)
 		err := scheduler.client.MultiRemove(triggerTaskKeys)
+		// after query coord restart, it will redo the triggerTask
+		// after recover, if child Task's state is TaskDone, will not be repeatedly executed
 		if err != nil {
+			log.Error("scheduleLoop: error when remove trigger task from etcd", zap.Int64("triggerTaskID", triggerTask.getTaskID()), zap.Error(err))
 			panic(err)
 		}
-		return nil
 	}
 
 	for {
 		var err error
 		select {
 		case <-scheduler.ctx.Done():
-			scheduler.stopActivateTaskLoopChan <- 1
 			return
 		case <-scheduler.triggerTaskQueue.Chan():
+			hasNotify := false
 			triggerTask = scheduler.triggerTaskQueue.popTask()
 			log.Debug("scheduleLoop: pop a triggerTask from triggerTaskQueue", zap.Int64("triggerTaskID", triggerTask.getTaskID()))
-			alreadyNotify := true
+			// the persistent task state is only taskUndo, taskDoing, taskDone
+			// taskUndo and taskDoing indicate that the assignment of internTasks for triggerTask was not completed
+			// taskDone said it had successfully completed the internTasks allocation for triggerTask
 			if triggerTask.getState() == taskUndo || triggerTask.getState() == taskDoing {
+				// if err == nil, the state of the triggerTask is already taskDone
+				// if err != nil, set the state of he triggerTask to taskFailed
 				err = scheduler.processTask(triggerTask)
 				if err != nil {
 					log.Debug("scheduleLoop: process triggerTask failed", zap.Int64("triggerTaskID", triggerTask.getTaskID()), zap.Error(err))
-					alreadyNotify = false
+					triggerTask.setState(taskFailed)
+				} else {
+					if triggerTask.msgType() == commonpb.MsgType_LoadCollection || triggerTask.msgType() == commonpb.MsgType_LoadCollection {
```

Comment:
this should not be allowed.
Task scheduler should never check triggerTask type, otherwise the implementation become tricks.

we can have a notify function for task, and only implement  in MsgType_LoadCollection.

Also the or seems to be or the same thing

Reply:
I also agree with it, the task should be scheduled by priorities, not task type.